app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/gainorlossafterreliefs/GlarCalculatorEligibilityUtil.scala

```diff
@@ -0,0 +1,151 @@
+  extends FrontendController(cc)
+    with WithAuthAndSessionDataAction
+    with SessionUpdates
+    with Logging
+    with StartingToAmendToFillingOutReturnBehaviour {
+
+  val glarCalculatorFlag = servicesConfig.getBoolean("glar-calculator.enabled")
+
+  def withValuesForCalculatorWhereEligible(fillingOutReturn: FillingOutReturn)(
+    f: Option[CalculatedGlarBreakdown] => Future[Result]
+  )(implicit request: RequestWithSessionData[_]): Future[Result] = {
+
+    def isEligible(
+                    fillingOutReturn: FillingOutReturn,
+                    completeSingleDisposalTriageAnswers: CompleteSingleDisposalTriageAnswers,
+                    reliefDetailsAnswers: CompleteReliefDetailsAnswers
+                  )(f: Boolean => Future[Result]): Future[Result] =
+      if (!glarCalculatorFlag) {
+        f(false)
+      } else {
+        val eligibleAssetType              =
+          completeSingleDisposalTriageAnswers.assetType === Residential || completeSingleDisposalTriageAnswers.assetType === NonResidential
+
+        val noOtherReliefs                 = reliefDetailsAnswers.otherReliefs.isEmpty
+
+        val underPreviousReturnLimit       = fillingOutReturn.previousSentReturns
+          .getOrElse(PreviousReturnData(summaries = List.empty, previousYearToDate = None))
+          .summaries
+          .length <= 10
+
+        val isSelf                         = completeSingleDisposalTriageAnswers.individualUserType.contains(Self)
+
+        val currentReturnIsEligible = eligibleAssetType && noOtherReliefs && underPreviousReturnLimit && isSelf
+
+        if (currentReturnIsEligible) {
+          val submissionIdsOfPreviousReturns = fillingOutReturn.previousSentReturns
+            .map(e => e.summaries)
+            .getOrElse(List.empty)
+            .map(e => e.submissionId)
+          val results = submissionIdsOfPreviousReturns.map {
+            returnsService
+              .displayReturn(fillingOutReturn.subscribedDetails.cgtReference, _)
+              .fold(
+                _ => false,
+                r =>
+                  r.completeReturn.fold(
+                    _ => false,
+                    e =>
+                      e.fold(
+                        _ => false,
+                        e => {
+                          val assetTypeCheck   =
+                            e.triageAnswers.assetType === completeSingleDisposalTriageAnswers.assetType
+                          val otherReliefCheck = e.reliefDetails.otherReliefs.isEmpty
+                          assetTypeCheck && otherReliefCheck
+                        },
+                        _ => false,
+                        _ => false,
+                        _ => false
+                      ),
+                    _ => false,
+                    _ => false,
+                    _ => false
+                  )
+              )
+          }
+
+          Future.sequence(results) flatMap { a =>
```

Comment:
you can sequence on an `EitherT` so you don't need to pop out the `EitherT` with the `fold` above

Reply:
I think you just need to import these:
```scala
import cats.syntax.traverse._
import cats.instances.future._
```
to be able to do something like:
```
val p: List[EitherT[Future, Error, X]] = ???
val q: EitherT[Future,Error,List[X]] = p.sequence[EitherT[Future,Error,*], X]
```